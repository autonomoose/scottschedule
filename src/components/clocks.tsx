// more clocks
//   - clocks must avoid hooks, clocks are conditionals
//   - define new clock and register it with widget

import React from 'react';

import Box from '@mui/material/Box';
import Button from '@mui/material/Button';
import Typography from '@mui/material/Typography';

// use this as an outside interface to clocks
interface ClockWidgetProps {
  currClock: string,
  onComplete: (status: string) => void,
  currSched?: string,
  descrSched?: string,
}
export const ClockWidget = (props: ClockWidgetProps) => {
    return(
      <Box display='flex' flexDirection='column'>
      {
        {
          'digital1': <ClockDigital1 onComplete={props.onComplete} />,
          'digital2': <ClockDigital2 onComplete={props.onComplete} />,
        }[props.currClock]
      }
      </Box>
)}

interface ClockProps {
  onComplete: (status: string) => void,
}
export const ClockDigital2 = (props: ClockProps) => {
    return(
      <>
        <Button data-testid='change clock' sx={{margin: 0}} onClick={() => props.onComplete('scheduler')}>
          <Typography variant='h1' id='mainclock' data-testid='mainclock'
            sx={{fontSize:150, fontWeight: 600, padding: 0, margin: 0}}>00:00</Typography>
        </Button>
        <Typography variant='subtitle1' id='mainpm' data-testid='mainpm' sx={{fontSize:20, fontWeight: 600}}>
          PM
        </Typography>
        <Box mx={4} display='flex' justifyContent='space-between'>
          <Typography mx={1} variant='h4' id='maindate' data-testid='maindate'>Day, 01/01</Typography>
          <Button onClick={() => props.onComplete('close')}>Scheduler</Button>
        </Box>
      </>
) };

export const ClockDigital1 = (props: ClockProps) => {
    return(
      <>
        <Button data-testid='change clock' sx={{margin: 0, padding: 0}} onClick={() => props.onComplete('digital2')}>
          <Box display='flex' justifyContent='flex-start' alignItems='stretch'>
            <Box mt={1} display='flex' alignItems='flex-start'>
              <Typography id='mainpm' data-testid='mainpm'
               sx={{lineHeight:1, fontSize:20, fontWeight: 600}}>
                PM
              </Typography>
            </Box>
            <Typography id='compclock' data-testid='compclock'
             sx={{marginLeft: '-2px', letterSpacing: '-30px', lineHeight:.8, fontSize:230, fontWeight: 600, padding: 0, margin: 0}}>
               0
            </Typography>
            <Box ml={3} mt={1}>
              <Typography id='compminutes' data-testid='compminutes'
               sx={{letterSpacing: '-4px',lineHeight:.92, fontSize:150, fontWeight: 800, padding: 0, margin: 0}}>
                00
              </Typography>
              <Typography id='maindate' data-testid='maindate'
               sx={{ textTransform: 'none', lineHeight:1, fontSize:26, padding: 0, margin: 0}}>
                Day, 01/01
              </Typography>
            </Box>
           </Box>
        </Button>
        <Box mx={4} display='flex' justifyContent='flex-end'>
          <Button onClick={() => props.onComplete('digital1-color')}>Color</Button>
          <Button onClick={() => props.onComplete('close')}>Scheduler</Button>
        </Box>
      </>
) };



