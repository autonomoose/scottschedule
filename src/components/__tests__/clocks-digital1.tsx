import React from "react";
import { render, waitFor } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
import "@testing-library/jest-dom";

import { ClockDigital1 } from "../clocks";

const mockCompleteFun = jest.fn();

// local test helpers
//
const mytest = <ClockDigital1 onComplete={mockCompleteFun}/>;
const mySetup = async () => {
    const utils = render(mytest);
    await waitFor(() => {
        expect(utils.getByTestId('change clock')).toBeVisible();
    });
    const chgClockButton = utils.getByTestId('change clock');
    const chgSchedButton = utils.getByRole('button', {name: /scheduler/i});
    const chgColorButton = utils.getByRole('button', {name: /color/i});
    return {
        ...utils,
        chgClockButton,
        chgSchedButton,
        chgColorButton,
    }
}

describe("ClockDigital1", () => {
  it("renders snapshot correctly", () => {
    const {asFragment} = render(mytest);
    expect(asFragment()).toMatchSnapshot();
  });

  it("changes clock when clock is clicked", async () => {
    const utils = await mySetup();

    userEvent.click(utils.chgClockButton);
    await waitFor(() => {
        expect(mockCompleteFun).toHaveBeenLastCalledWith('digital2');
    });
  });

  it("closes when scheduler is clicked", async () => {
    const utils = await mySetup();

    userEvent.click(utils.chgSchedButton);
    await waitFor(() => {
        expect(mockCompleteFun).toHaveBeenLastCalledWith('close');
    });
  });

  it("passes digital1-color when option is pressed", async () => {
    const utils = await mySetup();

    userEvent.click(utils.chgColorButton);
    await waitFor(() => {
        expect(mockCompleteFun).toHaveBeenLastCalledWith('digital1-color');
    });
  });

  it("defines clock update vars", async () => {
    const utils = await mySetup();

    expect(utils.getByTestId('mainpm')).toBeVisible();
    expect(utils.getByTestId('compclock')).toBeVisible();
    expect(utils.getByTestId('compminutes')).toBeVisible();
    expect(utils.getByTestId('maindate')).toBeVisible();
  });

});


