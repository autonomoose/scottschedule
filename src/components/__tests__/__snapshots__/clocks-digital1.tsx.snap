// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClockDigital1 renders snapshot correctly 1`] = `
<button
  class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium css-1oiecvs-MuiButtonBase-root-MuiButton-root"
  data-testid="change clock"
  tabindex="0"
  type="button"
>
  <div
    class="MuiBox-root css-1w0wdur"
  >
    <div
      class="MuiBox-root css-1mxzarq"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-bnbzw0-MuiTypography-root"
        data-testid="mainpm"
        id="mainpm"
      >
        PM
      </p>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body1 css-1qk3xk8-MuiTypography-root"
      data-testid="compclock"
      id="compclock"
    >
      0
    </p>
    <div
      class="MuiBox-root css-1b3ucee"
    >
      <p
        class="MuiTypography-root MuiTypography-body1 css-14da47l-MuiTypography-root"
        data-testid="compminutes"
        id="compminutes"
      >
        00
      </p>
      <p
        class="MuiTypography-root MuiTypography-body1 css-q6tqwp-MuiTypography-root"
        data-testid="maindate"
        id="maindate"
      >
        Day, 01/01
      </p>
    </div>
  </div>
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;
